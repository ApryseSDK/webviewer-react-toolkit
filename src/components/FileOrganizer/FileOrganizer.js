'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.FileOrganizer = void 0;
var tslib_1 = require('tslib');
var classnames_1 = tslib_1.__importDefault(require('classnames'));
var react_1 = tslib_1.__importStar(require('react'));
var utils_1 = require('../../utils');
var DndMultiProvider_1 = require('../DndMultiProvider');
var Draggable_1 = require('../Draggable');
var DragLayer_1 = require('../DragLayer');
var MemoAutoSizer_1 = require('./MemoAutoSizer');
var defaultSize = {
  width: utils_1.THUMBNAIL_WIDTH,
  height: utils_1.THUMBNAIL_WIDTH,
};
function FileOrganizer(_a) {
  var files = _a.files,
    onMove = _a.onMove,
    onDragChange = _a.onDragChange,
    onDeselectAll = _a.onDeselectAll,
    onSelectAll = _a.onSelectAll,
    onRenderThumbnail = _a.onRenderThumbnail,
    onRenderDragLayer = _a.onRenderDragLayer,
    disableMove = _a.disableMove,
    preventArrowsToMove = _a.preventArrowsToMove,
    preventClickAwayDeselect = _a.preventClickAwayDeselect,
    draggingIds = _a.draggingIds,
    padding = _a.padding,
    _gridRef = _a.gridRef,
    thumbnailSize = _a.thumbnailSize,
    noProvider = _a.noProvider,
    className = _a.className,
    onClick = _a.onClick,
    onKeyDown = _a.onKeyDown,
    style = _a.style,
    divProps = tslib_1.__rest(_a, [
      'files',
      'onMove',
      'onDragChange',
      'onDeselectAll',
      'onSelectAll',
      'onRenderThumbnail',
      'onRenderDragLayer',
      'disableMove',
      'preventArrowsToMove',
      'preventClickAwayDeselect',
      'draggingIds',
      'padding',
      'gridRef',
      'thumbnailSize',
      'noProvider',
      'className',
      'onClick',
      'onKeyDown',
      'style',
    ]);
  var fileOrganizerRef = (0, react_1.useRef)(null);
  var gridRef = (0, react_1.useRef)(null);
  (0, react_1.useImperativeHandle)(_gridRef, function() {
    return gridRef.current;
  });
  var _b = (0, react_1.useState)(0),
    columnCount = _b[0],
    setColumnCount = _b[1];
  var _c = (0, react_1.useState)(),
    editingId = _c[0],
    setEditingId = _c[1];
  var _d = (0, react_1.useState)(),
    draggingId = _d[0],
    setDraggingId = _d[1];
  // Get the width of the first item, or default if no first item found.
  var hasFiles = files.length > 0;
  var getSize = (0, react_1.useCallback)(
    function() {
      if (!hasFiles) return defaultSize;
      if (!fileOrganizerRef.current) return defaultSize;
      var draggableWrapper = fileOrganizerRef.current.querySelector(
        'div[draggable="true"]',
      );
      var draggableElement =
        draggableWrapper === null || draggableWrapper === void 0
          ? void 0
          : draggableWrapper.firstChild;
      var firstItem =
        draggableElement === null || draggableElement === void 0
          ? void 0
          : draggableElement.firstChild;
      if (!firstItem) return defaultSize;
      return firstItem.getBoundingClientRect();
    },
    [hasFiles],
  );
  // Detect size of first item and use as size throughout.
  var _e = (0, react_1.useState)(function() {
      return thumbnailSize || getSize();
    }),
    size = _e[0],
    setSize = _e[1];
  // Update size when getWidth ref changes (when hasFiles changes).
  (0, react_1.useEffect)(
    function() {
      if (thumbnailSize) return setSize(thumbnailSize);
      if (files.length === 0) return setSize(defaultSize);
      setSize(function(prev) {
        var _a = getSize(),
          width = _a.width,
          height = _a.height;
        if (prev.width === width && prev.height === height) return prev;
        return { width: width, height: height };
      });
      // Watches all files to continuously check width and height.
    },
    [files, getSize, thumbnailSize],
  );
  var handleOnDragChange = (0, react_1.useCallback)(
    function(id) {
      onDragChange === null || onDragChange === void 0
        ? void 0
        : onDragChange(id);
      setDraggingId(id);
    },
    [onDragChange],
  );
  var handleItemKeyDown = (0, react_1.useCallback)(
    function(event, index, _file, draggableRef) {
      var indexDiff = 0;
      switch (event.key) {
        case 'ArrowLeft':
          indexDiff = -1;
          break;
        case 'ArrowRight':
          indexDiff = 1;
          break;
        case 'ArrowUp':
          indexDiff = -1 * columnCount;
          break;
        case 'ArrowDown':
          indexDiff = columnCount;
          break;
        default:
          return; // Return if not one of above keys
      }
      event.preventDefault();
      var hasMoved = false;
      // If meta key was pressed, move to new location.
      if (
        !preventArrowsToMove &&
        (event.metaKey || event.ctrlKey) &&
        !disableMove &&
        editingId === undefined &&
        onMove
      ) {
        hasMoved = true;
        onMove(index, index + indexDiff);
      }
      if (!gridRef.current) return;
      var siblingAtLocation = (0, utils_1.getSibling)(
        draggableRef.current,
        indexDiff,
      );
      // If no meta key was pressed, focus item in direction of keys.
      if (siblingAtLocation && !(event.metaKey || event.ctrlKey)) {
        var focusable_1 = siblingAtLocation.querySelector(
          utils_1.focusableElementDomString,
        );
        if (focusable_1) {
          hasMoved = true;
          requestAnimationFrame(function() {
            focusable_1.focus();
          });
        }
      }
      if (!hasMoved) return;
      var isVisible = (0, utils_1.isScrolledIntoView)(
        siblingAtLocation,
        fileOrganizerRef.current,
      ).isVisible;
      if (isVisible) return;
      // Use react-window scrollToItem api for virtualized items.
      var rowIndex = (0, utils_1.getRowAndColumnIndex)(
        index + indexDiff,
        columnCount,
      ).rowIndex;
      gridRef.current.scrollToItem({ align: 'smart', rowIndex: rowIndex });
    },
    [columnCount, disableMove, editingId, onMove, preventArrowsToMove],
  );
  var pad = Math.max(1, padding || 0);
  var renderItem = (0, react_1.useCallback)(
    function(file, index, style) {
      if (!file)
        return react_1.default.createElement(
          react_1.Fragment,
          { key: '__null' },
          null,
        );
      var isEditing = editingId === file.id;
      var otherDragging = !!(
        (draggingId && draggingId !== file.id) ||
        (draggingIds && draggingIds.length && !draggingIds.includes(file.id))
      );
      var draggableRef = (0, react_1.createRef)();
      var isInDragGroup = draggingIds ? draggingIds.includes(file.id) : false;
      return react_1.default.createElement(Draggable_1.Draggable, {
        'data-file-id': file.id,
        key: file.id,
        index: index,
        style:
          style &&
          tslib_1.__assign(tslib_1.__assign({}, style), {
            top: typeof style.top === 'number' ? style.top + pad : style.top,
            left:
              typeof style.left === 'number' ? style.left + pad : style.left,
          }),
        ref: draggableRef,
        hideDragPreview: !!onRenderDragLayer,
        preventAnimation: isInDragGroup,
        onDragChange: function(isDragging) {
          return handleOnDragChange(isDragging ? file.id : undefined);
        },
        disableDrag: isEditing || disableMove,
        onMove: onMove,
        onKeyDown: function(e) {
          return handleItemKeyDown(e, index, file, draggableRef);
        },
        onRenderChildren: function(isDragging) {
          return onRenderThumbnail({
            onRenderThumbnailProps: {
              file: file,
              dragging: isDragging || isInDragGroup,
              otherDragging: otherDragging,
              onEditingChange: function(editing) {
                return setEditingId(editing ? file.id : undefined);
              },
            },
            id: file.id,
            index: index,
          });
        },
      });
    },
    [
      editingId,
      draggingId,
      draggingIds,
      pad,
      onRenderDragLayer,
      disableMove,
      onMove,
      handleOnDragChange,
      handleItemKeyDown,
      onRenderThumbnail,
    ],
  );
  var handleOnClick = (0, react_1.useCallback)(
    function(event) {
      onClick === null || onClick === void 0 ? void 0 : onClick(event);
      if (!preventClickAwayDeselect)
        onDeselectAll === null || onDeselectAll === void 0
          ? void 0
          : onDeselectAll();
    },
    [onDeselectAll, onClick, preventClickAwayDeselect],
  );
  var handleOnKeyDown = (0, react_1.useCallback)(
    function(event) {
      onKeyDown === null || onKeyDown === void 0 ? void 0 : onKeyDown(event);
      if (event.key === 'Escape')
        return onDeselectAll === null || onDeselectAll === void 0
          ? void 0
          : onDeselectAll();
      if (event.key === 'a' && (event.metaKey || event.ctrlKey)) {
        onSelectAll === null || onSelectAll === void 0 ? void 0 : onSelectAll();
        event.preventDefault();
      }
    },
    [onDeselectAll, onSelectAll, onKeyDown],
  );
  var customDragLayerTranslate = (0, react_1.useCallback)(function(_a) {
    var mousePosition = _a.mousePosition;
    var x = mousePosition.x - utils_1.THUMBNAIL_WIDTH / 2;
    var y = mousePosition.y - utils_1.THUMBNAIL_WIDTH / 2;
    return { x: x, y: y };
  }, []);
  var fileOrganizerClass = (0, classnames_1.default)(
    'ui__base ui__fileOrganizer',
    className,
  );
  var Wrapper = noProvider
    ? react_1.default.Fragment
    : DndMultiProvider_1.DndMultiProvider;
  return react_1.default.createElement(
    Wrapper,
    null,
    react_1.default.createElement(
      'div',
      tslib_1.__assign({}, divProps, {
        className: fileOrganizerClass,
        ref: fileOrganizerRef,
        onClick: handleOnClick,
        onKeyDown: handleOnKeyDown,
        style: style,
        role: 'grid',
        tabIndex: 0,
      }),
      react_1.default.createElement(MemoAutoSizer_1.MemoAutoSizer, {
        ref: gridRef,
        files: files,
        padding: pad,
        size: size,
        renderItem: renderItem,
        onColumnCountChange: setColumnCount,
      }),
      onRenderDragLayer
        ? react_1.default.createElement(
            DragLayer_1.DragLayer,
            { customTranslate: customDragLayerTranslate },
            react_1.default.createElement(
              'div',
              {
                className: 'ui__fileOrganizer__draglayer',
                style: {
                  height: utils_1.THUMBNAIL_WIDTH,
                  width: utils_1.THUMBNAIL_WIDTH,
                },
              },
              onRenderDragLayer(),
            ),
          )
        : undefined,
    ),
  );
}
exports.FileOrganizer = FileOrganizer;
//# sourceMappingURL=FileOrganizer.js.map
