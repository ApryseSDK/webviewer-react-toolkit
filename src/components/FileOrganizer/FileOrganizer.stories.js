'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.UseManagedFilesHook = exports.WithGridRef = exports.DifferentThumbnailSize = exports.WithCustomDragLayer = exports.EditableWithButtons = exports.Basic = void 0;
var tslib_1 = require('tslib');
var addon_knobs_1 = require('@storybook/addon-knobs');
var react_1 = tslib_1.__importStar(require('react'));
var hooks_1 = require('../../hooks');
var action_1 = require('../../storybook-helpers/action/action');
var files_1 = require('../../storybook-helpers/data/files');
var forwardAction_1 = require('../../storybook-helpers/knobs/forwardAction');
var integer_1 = require('../../storybook-helpers/knobs/integer');
var Button_1 = require('../Button');
var FileOrganizer_1 = require('../FileOrganizer');
var Icon_1 = require('../Icon');
var Thumbnail_1 = require('../Thumbnail');
var ThumbnailDragLayer_1 = require('../ThumbnailDragLayer');
var README_md_1 = tslib_1.__importDefault(require('./README.md'));
exports.default = {
  title: 'Components/FileOrganizer',
  component: FileOrganizer_1.FileOrganizer,
  parameters: { readme: README_md_1.default, info: { disable: true } },
};
var Template = function(_a) {
  var onRenderDragLayer = _a.onRenderDragLayer,
    _b = _a.numFiles,
    numFiles = _b === void 0 ? 2 : _b,
    editable = _a.editable,
    scrollToTop = _a.scrollToTop,
    customSizedThumbnail = _a.customSizedThumbnail;
  // This is the index organizing function.
  var _c = (0, react_1.useState)(function() {
      var newFiles = [];
      for (var index = 0; index < numFiles; index++) {
        newFiles.push((0, files_1.createFile)(index));
      }
      return newFiles;
    }),
    files = _c[0],
    setFiles = _c[1];
  var handleOnMove = (0, react_1.useCallback)(function(fromIndex, toIndex) {
    setFiles(function(prev) {
      if (toIndex < 0 || toIndex >= prev.length) return prev;
      var clone = prev.slice();
      var item = clone.splice(fromIndex, 1)[0];
      clone.splice(toIndex, 0, item);
      return clone;
    });
    return true;
  }, []);
  var gridRef = (0, react_1.useRef)(null);
  var onScrollToTop = function() {
    var _a;
    (_a = gridRef.current) === null || _a === void 0
      ? void 0
      : _a.scrollTo({ scrollTop: 0, scrollLeft: 0 });
  };
  var _d = (0, react_1.useState)(false),
    largerSize = _d[0],
    setLargerSize = _d[1];
  var changeSize = function() {
    return setLargerSize(function(prev) {
      return !prev;
    });
  };
  var size = largerSize
    ? { width: 200, height: 250 }
    : { width: 150, height: 200 };
  // This is just a helper for adding or removing files.
  (0, react_1.useEffect)(
    function() {
      setFiles(function(prev) {
        if (prev.length > numFiles) {
          return prev.slice(0, numFiles);
        }
        if (prev.length < numFiles) {
          var newFiles = [];
          for (var index = prev.length; index < numFiles; index++) {
            newFiles.push((0, files_1.createFile)(index));
          }
          return tslib_1.__spreadArray(
            tslib_1.__spreadArray([], prev, true),
            newFiles,
            true,
          );
        }
        return prev;
      });
    },
    [numFiles],
  );
  return react_1.default.createElement(
    react_1.default.Fragment,
    null,
    react_1.default.createElement(FileOrganizer_1.FileOrganizer, {
      files: files,
      gridRef: gridRef,
      preventArrowsToMove: (0, addon_knobs_1.boolean)(
        'preventArrowsToMove',
        false,
      ),
      preventClickAwayDeselect: (0, addon_knobs_1.boolean)(
        'preventClickAwayDeselect',
        false,
      ),
      disableMove: (0, addon_knobs_1.boolean)('disableMove', false),
      padding: (0, integer_1.integer)('padding', 0),
      onMove: (0, forwardAction_1.forwardAction)('onMove', handleOnMove),
      onDragChange: (0, action_1.action)('onDragChange'),
      onDeselectAll: (0, action_1.action)('onDeselectAll'),
      onSelectAll: (0, action_1.action)('onSelectAll'),
      thumbnailSize: customSizedThumbnail ? size : undefined,
      onRenderDragLayer: onRenderDragLayer
        ? function() {
            return react_1.default.createElement(
              ThumbnailDragLayer_1.ThumbnailDragLayer,
              null,
            );
          }
        : undefined,
      onRenderThumbnail: function(_a) {
        var onRenderThumbnailProps = _a.onRenderThumbnailProps,
          index = _a.index;
        return customSizedThumbnail
          ? react_1.default.createElement(
              'div',
              {
                style: tslib_1.__assign(tslib_1.__assign({}, size), {
                  display: 'flex',
                  justifyContent: 'center',
                  alignItems: 'center',
                  background: 'lightblue',
                  border: '1px solid orange',
                }),
              },
              'Thumbnail ',
              index + 1,
            )
          : react_1.default.createElement(
              Thumbnail_1.Thumbnail,
              tslib_1.__assign({}, onRenderThumbnailProps, {
                onRename: editable ? function() {} : undefined,
                buttonProps: editable
                  ? [
                      {
                        children: react_1.default.createElement(Icon_1.Icon, {
                          icon: 'Close',
                        }),
                        onClick: function() {},
                        key: 0,
                      },
                    ]
                  : undefined,
              }),
            );
      },
    }),
    scrollToTop &&
      react_1.default.createElement(
        Button_1.Button,
        { onClick: onScrollToTop },
        'Scroll to top',
      ),
    customSizedThumbnail &&
      react_1.default.createElement(
        Button_1.Button,
        { onClick: changeSize },
        'Change thumbnail size',
      ),
  );
};
var Basic = function() {
  var numFiles = (0, addon_knobs_1.number)('number of files', 2, {
    min: 0,
    max: 16,
    step: 1,
    range: true,
  });
  return react_1.default.createElement(Template, { numFiles: numFiles });
};
exports.Basic = Basic;
var EditableWithButtons = function() {
  var numFiles = (0, addon_knobs_1.number)('number of files', 2, {
    min: 0,
    max: 16,
    step: 1,
    range: true,
  });
  return react_1.default.createElement(Template, {
    numFiles: numFiles,
    editable: true,
  });
};
exports.EditableWithButtons = EditableWithButtons;
var WithCustomDragLayer = function() {
  var numFiles = (0, addon_knobs_1.number)('number of files', 2, {
    min: 0,
    max: 16,
    step: 1,
    range: true,
  });
  return react_1.default.createElement(Template, {
    numFiles: numFiles,
    onRenderDragLayer: true,
  });
};
exports.WithCustomDragLayer = WithCustomDragLayer;
var DifferentThumbnailSize = function() {
  var numFiles = (0, addon_knobs_1.number)('number of files', 2, {
    min: 0,
    max: 16,
    step: 1,
    range: true,
  });
  return react_1.default.createElement(Template, {
    numFiles: numFiles,
    customSizedThumbnail: true,
  });
};
exports.DifferentThumbnailSize = DifferentThumbnailSize;
var WithGridRef = function() {
  return react_1.default.createElement(Template, {
    numFiles: 100,
    scrollToTop: true,
  });
};
exports.WithGridRef = WithGridRef;
var UseManagedFilesHook = function() {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  var _a = (0, hooks_1.useManagedFiles)({
      initialFiles: Array.from({ length: 1000 }, function(_, index) {
        return (0, files_1.createFile)(index);
      }),
      preventMultiDrag: (0, addon_knobs_1.boolean)(
        'preventMultiDrag',
        false,
        'useManagedFiles options',
      ),
      preventMultiSelect: (0, addon_knobs_1.boolean)(
        'preventMultiSelect',
        false,
        'useManagedFiles options',
      ),
      preventDeselectOnDragOther: (0, addon_knobs_1.boolean)(
        'preventDeselectOnDragOther',
        false,
        'useManagedFiles options',
      ),
      preventSelectOnDrag: (0, addon_knobs_1.boolean)(
        'preventSelectOnDrag',
        false,
        'useManagedFiles options',
      ),
      easyMultiSelect: (0, addon_knobs_1.boolean)(
        'easyMultiSelect',
        false,
        'useManagedFiles options',
      ),
    }),
    fileOrganizerProps = _a.fileOrganizerProps,
    getThumbnailSelectionProps = _a.getThumbnailSelectionProps,
    draggingIds = _a.draggingIds;
  return react_1.default.createElement(
    FileOrganizer_1.FileOrganizer,
    tslib_1.__assign({}, fileOrganizerProps, {
      onRenderDragLayer: function() {
        return react_1.default.createElement(
          ThumbnailDragLayer_1.ThumbnailDragLayer,
          { numFiles: draggingIds.length },
        );
      },
      preventArrowsToMove: (0, addon_knobs_1.boolean)(
        'preventArrowsToMove',
        false,
        'FileOrganizer',
      ),
      preventClickAwayDeselect: (0, addon_knobs_1.boolean)(
        'preventClickAwayDeselect',
        false,
        'FileOrganizer',
      ),
      disableMove: (0, addon_knobs_1.boolean)(
        'disableMove',
        false,
        'FileOrganizer',
      ),
      padding: (0, integer_1.integer)('padding', 0, 'FileOrganizer'),
      onRenderThumbnail: function(_a) {
        var id = _a.id,
          onRenderThumbnailProps = _a.onRenderThumbnailProps;
        return react_1.default.createElement(
          Thumbnail_1.Thumbnail,
          tslib_1.__assign(
            {},
            getThumbnailSelectionProps(id),
            onRenderThumbnailProps,
          ),
        );
      },
    }),
  );
};
exports.UseManagedFilesHook = UseManagedFilesHook;
exports.UseManagedFilesHook.story = { name: 'useManagedFiles Hook' };
//# sourceMappingURL=FileOrganizer.stories.js.map
