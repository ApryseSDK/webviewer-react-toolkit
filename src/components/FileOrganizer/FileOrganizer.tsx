import classnames from 'classnames';
import React, { FC, KeyboardEvent, ReactNode, useState } from 'react';
import { DndProvider } from 'react-dnd';
import Backend from 'react-dnd-html5-backend';
import { File } from '../../data/file';
import Draggable from '../Draggable';
import DragLayer from '../DragLayer';

export interface FileOrganizerProps {
  /**
   * A list of files to render out within the page organizer.
   */
  files: File[];
  /**
   * On render function for generating the thumbnails for the page organizer.
   * If you do not want to build your own, try using the `Thumbnail` component.
   */
  onRenderThumbnail: (onRenderProps: OnRenderThumbnailProps) => ReactNode;
  /**
   * If provided, will call to render a custom drag layer while a thumbnail is
   * being dragged. Otherwise will show a preview of the thumbnail.
   */
  onRenderDragLayer?: () => ReactNode;
  /**
   * If true, will disable drag-and-drop functionality within the organizer.
   */
  disableMove?: boolean;
  /**
   * Callback fired when a file is moved within the page organizer.
   */
  onMove?: (fromIndex: number, toIndex: number, file: File) => void;
  /**
   * Classname for outer div.
   */
  className?: string;
  /**
   * Removes the ability to change indexes with arrow keys. This removes the
   * ability to re-sort accessibly. Generally, left and right arrow keys will
   * call `onMove` if a file is focused.
   */
  preventArrowsToMove?: boolean;
}

export interface OnRenderThumbnailProps {
  /** The file to render into a thumbnail. */
  file: File;
  /** The index of this file within the file organizer. */
  index: number;
  /** Is this file being dragged currently. */
  isDragging: boolean;
  /** Are other files being dragged other than this one. */
  otherDragging: boolean;
  /** Is this thumbnail being edited (dependent on you setting `setEditing`). */
  isEditing: boolean;
  /** Are other files being edited other than this one. */
  otherEditing: boolean;
  /** Callback for setting whether the thumbnail is in editing mode. */
  onEditingChange: (isEditing: boolean) => void;
}

export const FileOrganizer: FC<FileOrganizerProps> = ({
  files,
  onMove,
  onRenderThumbnail,
  onRenderDragLayer,
  disableMove,
  className,
  preventArrowsToMove,
}) => {
  const [editingList, setEditingList] = useState<string[]>([]);
  const setEditing = (id: string, isEditing: boolean) => {
    setEditingList(prev => {
      const prevIndex = prev.indexOf(id);
      if (isEditing && prevIndex === -1) return [...prev, id];
      if (!isEditing && prevIndex !== -1) return [...prev.slice(0, prevIndex), ...prev.slice(prevIndex + 1)];
      return prev;
    });
  };

  const [draggingList, setDraggingList] = useState<string[]>([]);
  const setDragging = (id: string, isDragging: boolean) => {
    setDraggingList(prev => {
      const prevIndex = prev.indexOf(id);
      if (isDragging && prevIndex === -1) return [...prev, id];
      if (!isDragging && prevIndex !== -1) return [...prev.slice(0, prevIndex), ...prev.slice(prevIndex + 1)];
      return prev;
    });
  };

  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>, index: number, file: File) => {
    if (preventArrowsToMove) return;
    if (editingList.includes(file.id)) return;
    if (event.key === 'ArrowLeft') {
      if (index === 0) return;
      onMove?.(index, index - 1, file);
      event.preventDefault();
    } else if (event.key === 'ArrowRight') {
      if (index === files.length - 1) return;
      onMove?.(index, index + 1, file);
      event.preventDefault();
    }
  };

  const fileOrganizerClass = classnames('ui__base ui__fileOrganizer', className);

  return (
    <DndProvider backend={Backend}>
      <div className={fileOrganizerClass}>
        {files.map((file, index) => {
          const isEditing = editingList.includes(file.id);
          const otherEditing = editingList.length > 0 && !editingList.includes(file.id);
          const otherDragging = draggingList.length > 0 && !draggingList.includes(file.id);
          return (
            <Draggable
              key={file.id}
              index={index}
              hideDragPreview={!!onRenderDragLayer}
              onDragChange={dragging => setDragging(file.id, dragging)}
              disableDrag={isEditing || disableMove}
              onMove={(fromIndex, toIndex) => onMove?.(fromIndex, toIndex, file)}
              onKeyDown={e => handleKeyDown(e, index, file)}
              onRenderChildren={isDragging => {
                return onRenderThumbnail({
                  file,
                  isDragging,
                  otherDragging,
                  isEditing,
                  otherEditing,
                  onEditingChange: editing => setEditing(file.id, editing),
                  index,
                });
              }}
            />
          );
        })}
        {onRenderDragLayer ? <DragLayer>{onRenderDragLayer()}</DragLayer> : null}
      </div>
    </DndProvider>
  );
};
