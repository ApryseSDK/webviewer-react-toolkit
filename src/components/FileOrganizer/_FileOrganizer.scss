.ui__fileOrganizer {
  background-color: $background-canvas;
  display: flex;
  flex-wrap: wrap;
}
