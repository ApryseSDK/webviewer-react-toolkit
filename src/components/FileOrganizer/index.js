'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var tslib_1 = require('tslib');
tslib_1.__exportStar(require('./FileOrganizer'), exports);
//# sourceMappingURL=index.js.map
