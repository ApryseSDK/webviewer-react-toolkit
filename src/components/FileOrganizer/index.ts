export { FileOrganizer as default, FileOrganizerProps } from './FileOrganizer';
