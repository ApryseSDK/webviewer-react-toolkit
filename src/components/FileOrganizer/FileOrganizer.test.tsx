import { shallow } from 'enzyme';
import React from 'react';
import FileOrganizer from '../FileOrganizer';
import { File } from '../../hooks/useFile';

const testFile: File = {
  id: 'test',
  name: 'Test name',
  extension: 'pdf',
  originalName: 'Test original name',
  mutateDocumentObj: () => console.log('test'),
  mutateFileObj: () => console.log('test'),
  setDocumentObj: async () => console.log('test'),
  setFileObj: async () => console.log('test'),
  setName: async () => console.log('test'),
};

describe('FileOrganizer component', () => {
  it('renders its contents', () => {
    const fileOrganizer = shallow(<FileOrganizer files={[testFile]} onRenderThumbnail={() => 'Thumbnail'} />);
    expect(fileOrganizer.find('.ui__fileOrganizer').length).toEqual(1);
  });

  it('snapshot renders default FileOrganizer', () => {
    const fileOrganizer = shallow(<FileOrganizer files={[testFile]} onRenderThumbnail={() => 'Thumbnail'} />);
    expect(fileOrganizer).toMatchSnapshot();
  });
});
