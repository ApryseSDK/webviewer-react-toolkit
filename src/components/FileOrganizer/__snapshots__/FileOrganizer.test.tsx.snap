// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileOrganizer component snapshot renders default FileOrganizer 1`] = `
<div
  className="ui__base ui__fileOrganizer"
>
  <DndProvider
    backend={[Function]}
  >
    <Component
      index={0}
      key="test"
      onKeyDown={[Function]}
      onMove={[Function]}
      onRenderChildren={[Function]}
    />
  </DndProvider>
</div>
`;
